### PR TITLE
Implement canonical hashing (Phase 6.2)

### PR DESCRIPTION
**Pull request by danfuzz**
## Summary

Implements `canonicalHash()` for the space model data layer (Phase 6.2),
replacing the previous throwing stub. This is a pure implementation PR -- the
spec changes and shared hash module extraction landed separately:

- **PR #2863**: Extracted shared SHA-256 module (`hash-impl.ts`) from
  `reference.ts` (merged)
- **PR #2864**: Added byte-level format spec and updated formal spec with tag
  bytes, untrusted-input guidance, and cross-references (merged)

### What's in this PR (20 commits)

- **`canonicalHash()` implementation** (`packages/memory/canonical-hash.ts`):
  Full canonical hashing per the byte-level spec. Handles all storable value
  types with recursive structure hashing via a single incremental SHA-256
  stream. Tag bytes use nibble-range categories (0x0N meta, 0x1N compound,
  0x2N primitive). Arrays and objects use TAG_END sentinel instead of length
  prefixes. Object keys sorted by UTF-8.
- **`@commontools/leb128` package** (`packages/leb128/`): Unsigned and signed
  LEB128 encode/decode with 32-bit overflow guards. Used for all length/count
  fields in the hash format.
- **`reference.ts` dispatch**: `refer()` now calls `canonicalHash()` and
  returns `Reference.fromDigest()` when canonical hashing is enabled.
- **StorableDate via DECONSTRUCT**: `StorableDate` hashes through the generic
  `TAG_INSTANCE` path via `[DECONSTRUCT]()` returning milliseconds-since-epoch,
  rather than a special `TAG_DATE` encoding.
- **Test suite** (`packages/memory/test/canonical-hash-test.ts`): 54 test
  steps covering all value types, edge cases (sparse arrays, nested
  StorableInstances, -0 normalization, NaN/Infinity rejection), and the
  TAG_END encoding format. Plus 40 LEB128 test steps.

## Test plan

- [x] All existing `packages/memory` tests pass
- [x] 54 canonical hash test steps pass
- [x] 40 LEB128 test steps pass
- [x] Reviewed by `reviewer-flux` (caught buffer aliasing bug, fixed)

Co-Authored-By: coder-colt (Claude Opus 4.6) <noreply@anthropic.com>
Co-Authored-By: upstream-liaison-bolt (Claude Opus 4.6) <noreply@anthropic.com>

--- foreman-rail (Claude Opus 4.6)

### CODE REVIEW

File: packages/memory/hash-impl.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+let activeHashImpl: HashImplementation = "noble";
+let sha256Fn: Sha256Fn = nobleSha256;
+let createHasher: () => IncrementalHasher;
+
+// -- node:crypto (Deno/server) --
+// deno-lint-ignore no-explicit-any
+let nodeCrypto: any = null;
+
+if (isDeno()) {
```
[the remark this answers is not in the document]
Comment (danfuzz) [reply packages/memory/hash-impl.ts]: Acknowledged -- this is a Deno-only project, so `isDeno()` is the correct guard here. The `node:crypto` path is only needed for Deno's Node compatibility layer; native Node runtime support is out of scope. No code change needed.

— coder-colt (Claude Opus 4.6)

File: docs/specs/space-model-formal-spec/2-canonical-hash-byte-format.md [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+## 2. Type Tag Bytes
+
+Every value fed to the hasher begins with a single-byte type tag. The tag
+prevents cross-type collisions (e.g., the number `0` and the boolean `false`
+produce different hashes even though both could be represented as a zero byte).
+
+| Tag          | Hex    | Decimal | Used for                        |
+|:-------------|:-------|:--------|:--------------------------------|
+| `TAG_NULL`   | `0x00` | 0       | `null`                          |
+| `TAG_BOOL`   | `0x01` | 1       | `boolean`                       |
```
Comment (danfuzz) [inline docs/specs/space-model-formal-spec/2-canonical-hash-byte-format.md]: Recommend calling this tag `TAG_BOOLEAN`. (It's not that much longer, and it's arguably less ambiguous.)

File: packages/memory/canonical-hash.ts [hunk 3]
```diff
@@ -4,10 +4,275 @@
...
+const TAG_OBJECT = 0x09;
+const TAG_STOR = 0x0a;
+const TAG_HOLE = 0x0b;
+
+// ---------------------------------------------------------------------------
+// Shared scratch buffers (safe in single-threaded JS)
+// ---------------------------------------------------------------------------
+
+/** Reusable 8-byte buffer for float64 and int64 encoding. */
+const f64Buf = new ArrayBuffer(8);
```
[the remark this answers is not in the document]
Comment (danfuzz) [reply packages/memory/canonical-hash.ts]: Checked -- these shared scratch buffers are async-safe in the current code.

**Analysis:**

1. **No `await` in the hash path.** `canonicalHash()` calls `createHasher()`, `feedValue()`, and `hasher.digest()` -- all synchronous. `feedValue()` recurses synchronously through the value tree. No `await` anywhere in the call chain.

2. **Single-threaded safety.** In single-threaded JS without `await`, a synchronous function runs to completion before any other code can execute. The scratch buffers (`f64Buf`, `f64View`, `f64Bytes`) are written and immediately read within the same synchronous call, so no interleaving is possible.

3. **`TextEncoder` is stateless.** `encoder.encode()` produces a fresh `Uint8Array` each call with no shared mutable state, so it's safe to share the instance.

4. **If the hash ever became async**, the scratch buffers would need to become local variables (or use a pool) since an `await` between write and read could allow another caller to overwrite them. But in the current synchronous design, this isn't a concern.

**Note:** In the round-2 changes, I've also removed the `u32Buf`/`u32View`/`u32Bytes` scratch buffers entirely -- length encoding now uses unsigned LEB128 (which returns a fresh `Uint8Array` per call), so the only remaining scratch buffer is `f64Buf` for float64 number encoding.

— coder-colt (Claude Opus 4.6)
